enterprise/suppress-user/factory.go

```diff
@@ -21,62 +25,216 @@ type Factory struct {
+		WithHttpClient(&http.Client{Timeout: config.GetDuration("HttpClient.suppressUser.timeout", 30, time.Second)}),
+		WithPageSize(config.GetInt("BackendConfig.Regulations.pageSize", 5000)),
+		WithPollIntervalFn(func() time.Duration { return pollInterval }),
+	)
+	if err != nil {
+		return nil, nil, err
+	}
+	return syncer, repo, nil
+}
+
 // Setup initializes the user suppression feature
 func (m *Factory) Setup(ctx context.Context, backendConfig backendconfig.BackendConfig) (types.UserSuppression, error) {
 	if m.Log == nil {
 		m.Log = logger.NewLogger().Child("enterprise").Child("suppress-user")
 	}
+
 	if m.EnterpriseToken == "" {
 		m.Log.Info("Suppress User feature is enterprise only")
 		return &NOOP{}, nil
 	}
+
 	m.Log.Info("Setting up Suppress User Feature")
+
 	backendConfig.WaitForConfig(ctx)
-	var repository Repository
-	if config.GetBool("BackendConfig.Regulations.useBadgerDB", true) {
-		tmpDir, err := misc.CreateTMPDIR()
+
+	var pollInterval time.Duration
+	config.RegisterDurationConfigVariable(300, &pollInterval, true, time.Second, "BackendConfig.Regulations.pollInterval")
+
+	useBadgerDB := config.GetBool("BackendConfig.Regulations.useBadgerDB", true)
+	if useBadgerDB {
+		identity := backendConfig.Identity()
+
+		fullSuppressionPath, latestSuppressionPath, err := getRepoPath()
 		if err != nil {
-			return nil, fmt.Errorf("could not create tmp dir: %w", err)
+			return nil, fmt.Errorf("could not get repo path: %w", err)
 		}
-		path := path.Join(tmpDir, "suppression")
 
-		// TODO: implement seeder source, to retrieve the initial state from a persisted backup
-		var seederSource func() (io.Reader, error)
+		if _, err = os.Stat(fullSuppressionPath); os.IsNotExist(err) && config.IsSet("SUPPRESS_USER_BACKUP_SERVICE_URL") {
+			repo := &RepoSwitcher{
+				mu: sync.RWMutex{},
+			}
 
-		repository, err = NewBadgerRepository(
-			path,
-			m.Log,
-			WithSeederSource(seederSource),
-			WithMaxSeedWait(config.GetDuration("BackendConfig.Regulations.maxSeedWait", 5, time.Second)))
-		if err != nil {
-			return nil, fmt.Errorf("could not create badger repository: %w", err)
+			latestSyncer, latestRepo, err := m.NewSyncerWithBadgerRepo(latestSuppressionPath, latestDataSeed, identity, pollInterval)
+			if err != nil {
+				return nil, err
+			}
+
+			subCtx, latestSyncCancel := context.WithCancel(ctx)
+			rruntime.Go(func() {
+				latestSyncer.SyncLoop(subCtx)
+				err = latestRepo.Stop()
+				if err != nil {
+					m.Log.Error("Latest Sync failed: could not stop repo: %w", err)
+				}
+				err = os.RemoveAll(latestSuppressionPath)
+				if err != nil {
+					m.Log.Error("Latest Sync failed: could not remove repo: %w", err)
+				}
+			})
+			repo.Repository = latestRepo
+			rruntime.Go(func() {
+				fullSyncer, fullRepo, err := m.NewSyncerWithBadgerRepo(fullSuppressionPath, fullDataSeed, identity, pollInterval)
+				if err != nil {
+					m.Log.Error("Complete Synce failed: could not create syncer: %w", err)
+					return
+				}
```

Comment:
This should not be performed asynchronously. If we either fail to create a `latestSyncer` or a `fullSyncer`, `Setup` must return with an error.

Reply:
I am not sure, how can we perform fullSync `synchronously` and still return handler after latest sync for gateway to start.